drivers/net/nsos_adapt.c

```diff
@@ -361,6 +361,25 @@ int nsos_adapt_bind(int fd, const struct nsos_mid_sockaddr *addr_mid, size_t add
 		return ret;
 	}
 
+/*
+ * Zephyr allows to bind two sockets separately to IPv4 and IPv6. In order to have that behavior on
+ * Linux set IPV6_V6ONLY socket option.
+ *
+ * See 'man 7 ipv6' for details.
+ */
+#ifdef IPPROTO_IPV6
+#ifdef IPV6_V6ONLY
+	if (addr->sa_family == AF_INET6) {
+		int set = 1;
+
+		ret = setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &set, sizeof(set));
+		if (ret < 0) {
+			nsi_print_warning("%s: failed to set IPV6_V6ONLY: %d", __func__, errno);
+		}
+	}
+#endif /* IPV6_V6ONLY */
+#endif /* IPPROTO_IPV6 */
```

Comment:
The functionality that can be controlled by `IPV6_V6ONLY` socket option is enabled by default, so there is by default need to create two sockets for IPv4/IPv6. Also if one disables `CONFIG_NET_IPV4_MAPPING_TO_IPV6` (which is `n` by default), then there is always need to create two sockets for IPv4 and IPv6.
So with this in mind, this can be simplified a bit to
```
if (IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6) &&
    addr->sa_family == AF_INET6) {
	int set = 1;

	ret = setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &set, sizeof(set));
	if (ret < 0) {
		nsi_print_warning("%s: failed to set IPV6_V6ONLY: %d", __func__, errno);
	}
}
```

Reply:
Okay, so my statements and assumptions in this commit were wrong. I was not aware (or even didn't check) that we have in Zephyr option for IPv6 to IPv4 mapping.

Your proposed changes make sense, with the exception that we should check above for `!IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6)` (negated condition). So if Zephyr has disabled IPv4->IPv6 mapping, then we will disable it in NSOS to have expected behavior. When disabling IPv6->IPv4 in NSOS with the code above (setting `IPV6_V6ONLY` flag on socket) will allow to bind two separate sockets, one for IPv4 and one for IPv6. That way we will achieve compatibility with `echo_server` sample, which will fix https://github.com/zephyrproject-rtos/zephyr/issues/71243. Does it make sense now?

Reply:
> Your proposed changes make sense, with the exception that we should check above for `!IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6)` (negated condition). So if Zephyr has disabled IPv4->IPv6 mapping, then we will disable it in NSOS to have expected behavior

No, that is not correct. The `IPV6_V6ONLY` is only used if `CONFIG_NET_IPV4_MAPPING_TO_IPV6` is enabled. The option tells that we can have IPv4 connection via IPv6 socket. If the option is not set (which is the default), then the `IPV6_V6ONLY` option is not used at all so we do not need to set it because one cannot have IPv4 connection via IPv6 socket in this case.

By default the `echo_server` does not enable the IPv4-mapping-to-IPv6 option, so user must have set it manually. Currently there are two network samples that set this option, `echo_service` and `echo`.
Regardless of whether the mapping is enabled or not, the commit message does not really tell **why** we would need to have this change, currently it is just stating that V6ONLY option is set but that can also seen from the commit itself.